File: java/BigNumbers2/src/lexer/Lexer.java
package lexer;

import token.Token;
import token.TokenKeywordConverter;
import token.TokenType;

public class Lexer {
    private String source;
    private char currentChar;
    private int position;
    private int readPosition;

    public Lexer() {
        position = 0;
        readPosition = 0;
    }

    public void load(String source) {
        this.source = source;
        position = 0;
        readPosition = 0;
        readChar();
    }

    private void skipWhitespaces() {
        while (currentChar == ' ') readChar();
    }

    private void readChar() {
        if (readPosition >= source.length()) {
            currentChar = 0;
        } else {
            currentChar = source.charAt(readPosition);
        }
        position = readPosition;
        readPosition++;
    }

    private char peekChar() {
        if (readPosition >= source.length()) return 0;
        return source.charAt(readPosition);
    }

    private String readBigNumber() {
        int start = position;
        while(isDigit()) {
            readChar();
        }
        return source.substring(start, position);
    }

    private String readLiteral() {
        int start = position;
        while(isLetter()) {
            readChar();
        }
        while(isLetter() || isDigit()) {
            readChar();
        }
        return source.substring(start, position);
    }

    private boolean isDigit() {
        return '0' <= currentChar && currentChar <= '9';
    }

    private boolean isLetter() {
        return ('a' <= currentChar && currentChar <= 'z') || ('A' <= currentChar && currentChar <= 'Z') || currentChar == '_';
    }

    private boolean isFloatDot() {
        return currentChar == '.';
    }

    private String readIdentifier() {
        return null;
    }

    private TokenType lookupKeyword(String literal) {
        if (TokenKeywordConverter.checkKeyword(literal)) return TokenKeywordConverter.convertKeyword(literal);
        return TokenType.IDENT;
    }

    public Token nextToken() {
        skipWhitespaces();
        boolean readNextChar = true;
        Token result = new Token();

        switch (currentChar) {
            case 0:
                result.setTokenLiteral("");
                result.setType(TokenType.EOF);
                break;
            case '+':
                result.setTokenLiteral("+");
                result.setType(TokenType.SUM);
                break;
            case '-':
                result.setTokenLiteral("-");
                result.setType(TokenType.MINUS);
                break;
            case '*':
                result.setTokenLiteral("*");
                result.setType(TokenType.MULTIPLY);
                break;
            case '/':
                result.setTokenLiteral("/");
                result.setType(TokenType.DIVIDE);
                break;
            default:
                if (isLetter()) {
                    readNextChar = false;
                    String literal = readLiteral();
                    result.setTokenLiteral(literal);
                    result.setType(lookupKeyword(literal));
                } else if (isDigit()) {
                    readNextChar = false;
                    String literal = readBigNumber();
                    result.setTokenLiteral(literal);
                    result.setType(TokenType.BIG_NUMBER);
                } else {
                    result.setTokenLiteral(String.valueOf(currentChar));
                    result.setType(TokenType.UNDEFINED);
                }
                break;
        }

        if (readNextChar) {
            readChar();
        }
        return result;
    }
}

File: cpp/repl/big-numbers2/lexer.cpp
#include "lexer.h"

Lexer::Lexer() {

}

Lexer::~Lexer() {

}

void Lexer::load(std::string s) {
    source = s;
    position = 0;
    readPosition = 0;
    readChar();
}

bool Lexer::isLetter() {
    return ('a' <= current_char && current_char <= 'z') || ('A' <= current_char && current_char <= 'Z')  || current_char == '_';
}

bool Lexer::isDigit() {
    return '0' <= current_char && current_char <= '9';
}

void Lexer::skipWhitespaces() {
    while(current_char == ' ') readChar();
}

std::string Lexer::readBigNumber() {
    int start = position;
    while(isDigit()) {
        readChar();
    }
    return source.substr(start, position - start);
}

std::string Lexer::readLiteral() {
    int start = position;
    while(isLetter()) {
        readChar();
    }
    while(isLetter() || isDigit()) {
        readChar();
    }
    return source.substr(start, position - start);
}

void Lexer::readChar() {
    if (readPosition >= source.length()) {
        current_char = 0;
    } else {
        current_char = source[readPosition];
    }
    position = readPosition;
    readPosition++;
}

TokenType Lexer::lookupKeyword(std::string literal) {
    if (keywords.find(literal) != keywords.end()) return keywords.find(literal)->second;
    return TT_IDENT;
}

Token Lexer::nextToken() {
    Token result;
    bool read_next_char = true;
    skipWhitespaces();
    switch(current_char) {
        case 0:
            result.setLiteral("");
            result.setType(TT_EOF);
        break;
        case '+':
            result.setLiteral("+");
            result.setType(TT_SUM);
        break;
        case '-':
            result.setLiteral("-");
            result.setType(TT_MINUS);
        break;
        case '*':
            result.setLiteral("*");
            result.setType(TT_MULTIPLY);
        break;
        case '/':
            result.setLiteral("/");
            result.setType(TT_DIVIDE);
        break;
        case '{':
            result.setLiteral("{");
            result.setType(TT_LBRACKET);
        break;
        case '}':
            result.setLiteral("}");
            result.setType(TT_RBRACKET);
        break;
        default:
            if (isLetter()) {
                read_next_char = false;
                std::string literal = readLiteral();
                result.setLiteral(literal);
                result.setType(lookupKeyword(literal));
            } else if (isDigit()) {
                read_next_char = false;
                std::string literal = readBigNumber();
                result.setLiteral(literal);
                result.setType(TT_BIGNUMBER);
            } else {
                std::string literal;
                literal.push_back(current_char);
                result.setLiteral(literal);
                result.setType(TT_ILLEGAL);
            }
        break;
    }
    if (read_next_char) readChar();
    return result;
}